### ISSUE TITLE
 AttributeError: 'CacheKey' object has no attribute 'encode'

### ISSUE THREAD

**Issue by cleder** [jamesls/fakeredis#129]
core/models.py:2535: in save
    return super(Topic, self).save(_args, *_kwargs)
core/models.py:293: in save
    return super(TimestampedBase, self).save(_args, *_kwargs)
../.virtualenvs/primarysite/local/lib/python2.7/site-packages/django/db/models/base.py:734: in save
    force_update=force_update, update_fields=update_fields)
../.virtualenvs/primarysite/local/lib/python2.7/site-packages/django/db/models/base.py:771: in save_base
    update_fields=update_fields, raw=raw, using=using)
../.virtualenvs/primarysite/local/lib/python2.7/site-packages/django/dispatch/dispatcher.py:189: in send
    response = receiver(signal=self, sender=sender, *_named)
core/signal_handlers/cache_signals.py:27: in invalidate_navigation_cache
    invalidate_all_topic_menu_node_cache(site_id=instance.site.id)
core/utils/caching.py:120: in invalidate_all_topic_menu_node_cache
    cache.delete_pattern(cache_key_pattern)
../.virtualenvs/primarysite/local/lib/python2.7/site-packages/django_redis/cache.py:32: in _decorator
    return method(self, *args, *_kwargs)
../.virtualenvs/primarysite/local/lib/python2.7/site-packages/django_redis/cache.py:95: in delete_pattern
    return self.client.delete_pattern(_args, *_kwargs)
../.virtualenvs/primarysite/local/lib/python2.7/site-packages/django_redis/client/default.py:262: in delete_pattern
    for key in client.scan_iter(pattern):
../.virtualenvs/primarysite/local/lib/python2.7/site-packages/fakeredis.py:1574: in scan_iter
    cursor, data = self.scan(cursor=cursor, match=match, count=count)
../.virtualenvs/primarysite/local/lib/python2.7/site-packages/fakeredis.py:1558: in scan
    return self._scan(self.keys(), int(cursor), match, count or 10)
../.virtualenvs/primarysite/local/lib/python2.7/site-packages/fakeredis.py:1551: in _scan
    if not match or fnmatch.fnmatch(to_native(val), to_native(match)):

---

x = <django_redis.util.CacheKey object at 0x8dd1090>, charset = 'ascii', errors = 'strict'

```
def to_native(x, charset=sys.getdefaultencoding(), errors='strict'):
    if x is None or isinstance(x, str):
        return x
```

> ```
>   return x.encode(charset, errors)
> ```
> 
> E       AttributeError: 'CacheKey' object has no attribute 'encode'

../.virtualenvs/primarysite/local/lib/python2.7/site-packages/fakeredis.py:56: AttributeError

**Comment by jamesls** [comment]
Could you share a code sample for how to repro this issue?

As your traceback shows, the `CacheKey` doesn't come from fakeredis, it's from django_redis, and looking at it's implementation here: https://github.com/niwinz/django-redis/blob/b8c6f6216bfea2f8a587efefda83ffa5208fb90d/django_redis/util.py#L24 it subclasses from `str`.  The code in question in fakeredis is:

```
    def to_native(x, charset=sys.getdefaultencoding(), errors='strict'):
        if x is None or isinstance(x, str):
            return x
        return x.encode(charset, errors)

```

So it should be hitting the `if` clause and not calling encode, which is what happens if I run this in the repl:

```
>>> from django_redis.util import CacheKey
>>> c = CacheKey('test')
>>> isinstance(c, str)
True
```

**Comment by kouk** [comment]
@jamesis this is the case in django_redis 4.5 and above, but for folks still supporting Django 1.6 who have to use django_redis 4.4.4 CacheKey is *not* a subclass of `str`: https://github.com/niwinz/django-redis/commit/739fc6e668ba7d7d2a2707b2e7444812a1836fe4#diff-87d115041b627dfcbc0e6fbf69fb70ba

**Comment by bmerry** [comment]
@cleder @kouk could one of you try out #184 (which eliminates to_native entirely) and let me know if it fixes the issue?

**Comment by kouk** [comment]
@bmerry yes, it fixes the issue for me, still using Django 1.6.  Thank you!
